src/MagnumPlugins/BasisImporter/BasisImporter.cpp

```diff
@@ -0,0 +1,217 @@
+    /* Does nothing if already initialized */
+    basist::basisu_transcoder_init();
+
+    /* Initializes codebook */
+    _state.reset(new State);
+
+    /* Set format configuration from plugin alias
+       or set a default */
+    if(Corrade::Utility::String::beginsWith(plugin, "BasisImporter")) {
+
+        /* Has type prefix */
+        if(plugin.length() > 13) {
+            /* We can assume the substring results in a valid value
+               as the plugin conf limits it to known suffixes */
+            configuration().setValue("format", plugin.substr(13));
+
+        /* Default configuration to Etc2 */
+        } else {
+            configuration().setValue("format",
+                BasisTranscodingType::Etc2);
+        }
+    }
+}
+
+BasisImporter::~BasisImporter() = default;
+
+auto BasisImporter::doFeatures() const -> Features { return Feature::OpenData; }
+
+bool BasisImporter::doIsOpened() const { return _in; }
+
+void BasisImporter::doClose() { _in = nullptr; }
+
+void BasisImporter::doOpenData(const Containers::ArrayView<const char> data) {
+    /* Because here we're copying the data and using the _in to check if file
+       is opened, having them nullptr would mean openData() would fail without
+       any error message. It's not possible to do this check on the importer
+       side, because empty file is valid in some formats (OBJ or glTF). We also
+       can't do the full import here because then doImage2D() would need to
+       copy the imported data instead anyway (and the uncompressed size is much
+       larger). This way it'll also work nicely with a future openMemory(). */
+    if(data.empty()) {
+        Error{} << "Trade::BasisImporter::openData(): the file is empty";
+        return;
+    }
+
+    if(!_state->_transcoder.validate_header(data.data(), data.size())) {
+        Error() << "Trade::BasisImporter::openData(): invalid header";
+        return;
+    }
+
+    if(!_state->_transcoder.start_transcoding(data.data(), data.size())) {
+        Error() << "Trade::BasisImporter::openData(): bad basis file";
+        return;
+    }
+
+    _in = Containers::Array<unsigned char>(data.size());
+    std::copy(data.begin(), data.end(), _in.begin());
+}
+
+UnsignedInt BasisImporter::doImage2DCount() const {
+    return _state->_transcoder.get_total_images(_in.data(), _in.size());
+}
+
+Containers::Optional<ImageData2D> BasisImporter::doImage2D(UnsignedInt index) {
+    const UnsignedInt level = 0;
+
+    const BasisTranscodingType targetFormat =
+        configuration().value<BasisTranscodingType>("format");
+    const basist::transcoder_texture_format format =
+        getTranscoderTextureFormat(targetFormat);
+    const UnsignedInt bytes_per_block = basis_get_bytes_per_block(format);
+
+    basist::basisu_image_info info;
+    if (!_state->_transcoder.get_image_info(_in.data(), _in.size(), info, index)) {
+        Error{} << "Trade::BasisImporter::doImage2D(): unabled to get image info";
```

Comment:
unable, not unabled :) Same below.

Besides that, the prefix should have the public-facing API name to avoid confusion, so `image2D()` instead of `doImage2D()`. Similarly in other cases.

Reply:
Ah, of course 😅